### PR TITLE
allow automatic simplification of sqrt if result can be written without sqrt

### LINKED ISSUE #12739: sqrt splits out non-real factors
```
>>> sqrt((3 + 4*I)/(3 - 4*I))
sqrt(-1/(3 - 4*I))*sqrt(-3 - 4*I)
```

It does this because that factor is nonnegative (but it's not real so it should remain in the sqrt).

I have this fixed in #12472; this is here as a reminder to make sure this is tested.

### PR DESCRIPTION
**Pull request by smichr**
`sqrt(a + I*b)` can sometimes be written simply as `c + I*d`. This PR checks for those conditions. Tests are added that demonstrate the new capability. The principle value of the `sqrt` is returned (see Algebraic Formula [here](https://en.wikipedia.org/w/index.php?title=Square_root)).

- [x] have all modifications recognize (a+I*b)**(n/2) where n is an integer and return the expanded result, e.g. `(3+4*I)**(S(3)/2)` is `2+11*I`
- [x] handle `sqrt(-2*I)` in a fashion similar to `sqrt(2*I)` except result will be `r - I*r`

closes #12739

### CODE REVIEW

**Comment by smichr** [comment]
@asmeurer , any objections or should this be part of `_simplify_basic` for power (or whatever the basic simplification method is)? Or maybe part of `re` and `im`? Right now, this sort of simplification is not recognized:

```
>>> sqrt(32 - 24*I)
sqrt(32 - 24*I)
>>> simplify(_)
2*sqrt(8 - 6*I)  # could be 6 - 2*I
>>> _.n()
6.0 - 2.0*I
```

**Comment by smichr** [comment]
@jksuom , do you have any feedback about this?

**Comment by jksuom** [comment]
I'm inclined to agree that some automatic simplifications like this would be acceptable and even useful. In general, there seems to be no reason for using an expression like `sqrt(3 + 4*I)` to denote a plain complex number with integer or rational components like `2 + I`. If that is really wanted, it could be constructed by setting `evaluate=False`.

This only affects a rather limited class of expressions. It seems to work also in some other cases like
```
>>> sqrt(1/(3 + 4*I))
2/5 - I/5
```
but not with fractions of type `(a + b*I)/(c + d*I)`:
```
>>> sqrt((3 + 4*I)/(3 - 4*I))
sqrt(-1/(3 - 4*I))*(1 - 2*I)
```
It looks like `1/(3 + 4*I)` is internally transformed to `(3 - 4*I)/25` at some point, but the corresponding transformation is not done on the fraction. It would be interesting to see if that could be arranged. Then this simplification would probably also work on fractions. But that need not be the subject of this PR.
Hence I am +1 for this.

File: sympy/core/add.py [hunk 1]
```diff
@@ -345,6 +345,25 @@ def as_coeff_Add(self, rational=False):
     # let Expr.as_coeff_mul() just always return (S.One, self) for an Add.  See
     # issue 5524.
 
+    def _eval_power(self, e):
+        if e is S.Half and self.is_number:
+            from sympy.core.evalf import pure_complex
+            from sympy.functions.elementary.complexes import sign
+            from sympy.functions.elementary.miscellaneous import sqrt
+            n, d = self.as_numer_denom()
```
Comment (asmeurer) [inline sympy/core/add.py]: What is the point of splitting out `d` here? Why not only operate on `pure_complex(self)`?

File: sympy/core/tests/test_eval_power.py [hunk 2]
```diff
@@ -347,5 +347,25 @@ def test_issue_8582():
 def test_issue_8650():
     n = Symbol('n', integer=True, nonnegative=True)
     assert (n**n).is_positive is True
-    x = 5*n+5
-    assert (x**(5*(n+1))).is_positive is True
+    x = 5*n + 5
+    assert (x**(5*(n + 1))).is_positive is True
+
+
+def test_better_sqrt():
```
Comment (asmeurer) [inline sympy/core/tests/test_eval_power.py]: Should add some tests for the cases where it doesn't simplify.

File: sympy/core/add.py (hunk 1 above)
Comment (smichr) [inline sympy/core/add.py]: I think I was trying to fail early. I just rewrote this. Let's see how the tests go.

**Comment by smichr** [comment]
> It would be interesting to see if that could be arranged

I have modified the `Pow._eval_power` routine to look for power `1/(a+I*b)` and the `Mul._eval_power` routine to look for `n/(a+I*b)`. Tests were added and all added lines are covered.

@asmeurer , I now only process `self` instead of the numerator and denominator.

I have tried to avoid the error of treating sqrt(a/b) as sqrt(a)/sqrt(b) except when `a` or `b` is positive.

**Comment by smichr** [comment]
ping @jksuom 
ping @asmeurer

File: sympy/core/tests/test_eval_power.py [hunk 3]
```diff
@@ -347,5 +347,41 @@ def test_issue_8582():
...
+    assert sqrt((3 + 4*I)/4) == 1 + I/2
+    # triple (8, 15, 17)
+    # parity of 8 doesn't match parity of 17 but
+    # den/2, 8/2, is a square
+    assert sqrt((8 + 15*I)/8) == (5 + 3*I)/4
+    # handle the denominator
+    assert sqrt((3 - 4*I)/25) == (2 - I)/5
+    assert sqrt((3 - 4*I)/26) == (2 - I)/sqrt(26)
+    # mul
+    assert sqrt((3 + 4*I)/(3 - 4*I)) == (1 - 2*I)/(1 + 2*I)  # not (2 + I)/(2 - I)
```
Comment (jksuom) [inline sympy/core/tests/test_eval_power.py]: `(1 - 2*I)/(1 + 2*I)` has negative real part.

**Comment by jksuom** [comment]
I wonder if there is any reason to leave the denominators in complex form. We could always transform `1/(a + b*I)` into `(a - b*I)/c` where `c = a**2 + b**2`. Then `sqrt((3 + 4*I)/(3 - 4*I))` would become

    (2 + I)/(2 - I) = (2 + I)**2/5 = 3/5 + 4/5*I

File: sympy/core/tests/test_eval_power.py (hunk 3 above)
Comment (smichr) [inline sympy/core/tests/test_eval_power.py]: :-( Hmmm. I'll have to see what happened with that.

Comment (smichr) [inline sympy/core/tests/test_eval_power.py]: This is a consequence of how master (incorrectly) splits up the arguments:

```
>>> sqrt((3 + 4*I)/(3 - 4*I))
sqrt(-1/(3 - 4*I))*sqrt(-3 - 4*I)
```

**Comment by smichr** [comment]
> I wonder if there is any reason to leave the denominators in complex form

One reason might be to keep numbers smaller. e.g.  with powers more negative than -1, the numbers can become large and esthetically I find that less appealing:

```
(3 + 4*I)/(3 - 4*I)**3 == (-117 + 44*I)/15625*(3+4*I)
```

But let me know what you think. 

What I've got only expands when the power is -1 but that can be easily changed (by reverting the last commit). That will also simplify the `_eval` code, too. I think what I have is too special cased right now. If the simple `a + I*b` in the denominator automatically comes to the numerator then the only routine that needs changing is the `Add._eval_power`: for the case of exponent of -1 and the case when the exponent has a denominator of 2.

File: sympy/core/tests/test_eval_power.py (hunk 3 above)
Comment (smichr) [inline sympy/core/tests/test_eval_power.py]: That is now fixed.

**Comment by smichr** [comment]
> Trouble finding roots #12775

ping @jksuom , that is now added, too.

Other things related to real and imaginary calculation broke when bringing `a+I*b` out of the denominator. I think I have those fixed. If the tests pass this is ready for a final review.

File: sympy/core/add.py [hunk 4]
```diff
@@ -350,6 +350,28 @@ def as_coeff_Add(self, rational=False):
...
+            from sympy.functions.elementary.miscellaneous import sqrt
+            ri = pure_complex(self)
+            if ri:
+                r, i = ri
+                if e.q == 2:
+                    D = sqrt(r**2 + i**2)
+                    if D.is_Rational:
+                        # (r, i, D) is a Pythagorean triple
+                        root = sqrt(factor_terms((D - r)/2))
+                        return root*expand_multinomial((
```
Comment (jksuom) [inline sympy/core/add.py]: Should we have also `root**e.p`?

Comment (smichr) [inline sympy/core/add.py]: I'll have to look later today. Maybe that is the source of troubles I am having with apart and the prudnikov test ([see here](https://travis-ci.org/sympy/sympy/jobs/245227050) for example).

**Comment by jksuom** [comment]
It seems that the failure is induced by the improved `.as_real_imag`:
```
>>> (-(3*2**pi)**(1/pi)).as_real_imag()
(-(3*2**pi)**(1/pi), 0)    # current master

>>> (-(3*2**pi)**(1/pi)).as_real_imag()
(-2*3**(1/pi), 0)    # this PR
```
This makes `d == 0` in `test_issue_10304` so that `d.is_comparable` is `True`.

**Comment by smichr** [comment]
> the failure is induced by the improved

That failure I worked around, but if you check out the link you will see a failure in hyperexpand. THAT's the sore spot right now.

**Comment by jksuom** [comment]
It seems that the failure in hyperexpand is the same as elsewhere, only it is invisible since bottom_up will catch the AttributeError.

**Comment by smichr** [comment]
> failure in hyperexpand

I tracked down the failure and now use `expand_mul` in the simplification which aims at simplifying a rational function. Basically, something like `cancel((1+I)*(1-I)/(2*x))` is being attempted and the polys created from the numerator and denominator don't cancel without an error. If `Poly` were used instead of `poly` this wouldn't be an issue. I left a note in the commit message titled "use expand_mul in carryout_plan".

I thought that modifying `together` would be a good idea (since simpler `together` expressions would result) but the current naive behavior is expected so I just left the extra expansion in the `partfrac` test. Otherwise an option `expand` would have to be given to `together`. I did that and reverted it as shown in the history.

**Comment by jksuom** [comment]
This looks like the old problem with polys over the EX domain: https://github.com/sympy/sympy/issues/12345, https://github.com/sympy/sympy/issues/5428. We never decided to enable extension=True by default, but in this case I suppose it could be set [here](https://github.com/sympy/sympy/blob/master/sympy/simplify/hyperexpand.py#L465) and probably also elsewhere in hyperexpand. That would make the simplification of expressions containing I automatic.

**Comment by smichr** [comment]
> That would make the simplification of expressions containing I automatic.

I'm not sure that I need simplification in general...the `expand_mul` is simplification enough in this case. (And I don't have the imagination to think of cases where this would not be enough.) Is this too much of a hack or does this look ok to you? I could also send the `expand=True` keyword as in `poly(numer, expand=True)` but I'm not sure if that's overkill for this need. Apparently the person writing this code thought `poly` instead of `Poly` was going to be sufficient so I don't want to bog things down with unnecessary steps.

**Comment by smichr** [comment]
> .the expand_mul is simplification enough

No, it's not. A few lines down another test failed. I'm trying the `expand=True` flag now.

**Comment by smichr** [comment]
It was the old "ball of string" with this one. I think all the issues have been resolved. This is ready for a final review.

**Comment by asmeurer** [comment]
extension=True could have some implications. It's not enabled by default for performance purposes.

**Comment by jksuom** [comment]
I have tested this with a version of `construct_domain` that yields `QQ<I>` instead of `EX` when there are complex coefficients but `extension=False` (essentially doing [this](https://github.com/sympy/sympy/blob/master/sympy/polys/constructor.py#L22)). It seems to work with no need to use `expand_mul` in `hyperexpand`.

**Comment by asmeurer** [comment]
Conditionally doing extension=True just for `I` (but not yet general algebraic numbers) is probably a safer change. Ideally it would always be True, but (unless this has changed) there are still performance issues if you do that with more than a couple of extensions.

**Comment by smichr** [comment]
The way it is right now, a full expand is needed on the numerator to pass all the `pudnikov` tests. @jksuom , do you want to do a PR to this or should this be committed as is?

**Comment by jksuom** [comment]
This looks good to me on the whole. I only feel a bit uneasy about the `expand` in hyperexpand since it seems unrelated to the code. It masks an issue that should be taken care of at lower level, for example in `construct_domain` the way I had tested. But it may be too early to merge my tentative PR before more extensive tests. I think this could be merged now as it is.

**Comment by smichr** [comment]
> . But it may be too early to merge my tentative PR

OK. I'll merge as is. Thanks for your help.

**Pull request merged**